### PR TITLE
Add headers for remote spec

### PR DESCRIPTION
**Pull request by Nexushunter**
With the next gen version of the generator now consumable it would be nice to provide the ability to add custom bearer tokens to the fetch spec requests. 

This adds the follow:
- A new base `InputSpec` type that represents local file access
- A `RemoteSpec` type that represents a hosted spec file.
  -  This is currently only supporting a bearer token and doesn't pull from the environment
    - This could be refined to look for something link OAS_BEARER or something instead of hard coding. I'm not sure yet.
  - An AWS specific `AwsRemoteSpec` that allows for access to specs in authenticated buckets
    -  Localstack for testing aws support

Attempts to solve: https://github.com/gibahjoe/openapi-generator-dart/issues/80

### CODE REVIEW

File: openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart [hunk 1]
```diff
@@ -148,6 +168,234 @@ class Openapi {
...
+      ...super.toJsonMap(),
+    };
+  }
+
+  AwsRemoteSpec.fromMap(Map<String, dynamic> map)
+      : bucket = map['bucket'],
+        region = map['region'],
+        _accessKeyId = map['accessKeyId'],
+        _secretAccessKey = map['secretAccessKey'],
+        now = DateTime.now(),
```
Comment (Nexushunter) [inline openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:371]: I'm likely going to set this as null to trigger toHeaderMap to reload the creds as `now` needs to be correct within up to 15 minutes of the request time

File: openapi-generator-annotations/pubspec.yaml [hunk 2]
```diff
@@ -5,7 +5,7 @@ homepage: https://github.com/gibahjoe/openapi-generator-dart
 
 
 environment:
-  sdk: '>=2.12.0 <3.0.0'
+  sdk: '>=2.17.0 <3.0.0'
```
Comment (Nexushunter) [inline openapi-generator-annotations/pubspec.yaml:8]: This allows for usage of `ConstructorName({super.field})`

File: openapi-generator-annotations/test/openapi_generator_annotations_test.dart [hunk 3]
```diff
@@ -28,30 +37,167 @@ void main() {
...
+                Platform.environment['AWS_SECRET_ACCESS_KEY'] = 'test';
+                final originalNow = remoteSpec.now;
+                final map = remoteSpec.toHeaderMap();
+                final newNow = remoteSpec.now;
+                Platform.environment.remove('AWS_ACCESS_KEY_ID');
+                Platform.environment.remove('AWS_SECRET_ACCESS_KEY');
+                expect(map['Authorization'], isNotNull);
+                expect(map['x-amz-date'], isNotNull);
+                expect(originalNow != newNow, isTrue);
+              }, skip: true);
```
Comment (Nexushunter) [inline openapi-generator-annotations/test/openapi_generator_annotations_test.dart:142]: I'm not used to working with the isolates and the docs kinda suck around usage will play around with it.

File: openapi-generator/lib/src/extensions/type_methods.dart
Comment (Nexushunter) [inline openapi-generator/lib/src/extensions/type_methods.dart:1]: Better type recognition here fixes some edge cases where the types could be coerced into invalid types.

File: openapi-generator/test/specs/next_gen_builder_test_aws_config.dart
Comment (Nexushunter) [inline openapi-generator/test/specs/next_gen_builder_test_aws_config.dart:1]: This may not be needed 🤷🏻

File: openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart [hunk 4]
```diff
@@ -148,6 +168,234 @@ class Openapi {
...
+///
+/// Since AWS handles the authentication header differently, we need to inform
+/// the builder to include the alternate auth header.
+///
+/// This currently only support AWS S3.
+///
+/// This contains authentication information for fetching the OAS spec ONLY. This
+/// does not apply security to the entry points defined in the OAS spec.
+class AwsRemoteSpec extends RemoteSpec {
+  /// The accessKeyId use to interact with AWS.
```
Comment (gibahjoe) [inline openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:247]: I think Remotespec should provide a ```header``` property instead of dealing with the implementation details of composing an auth header for AWS and other remote servers. This way, the consumer deals with the edge cases themselves (environments, etc). Then there will be no need of creating ```AwsRemoteSpec```,``` GCPRemoteSpec``` etc

File: openapi-generator-annotations/pubspec.yaml (hunk 2 above)
Comment (gibahjoe) [reply openapi-generator-annotations/pubspec.yaml:8]: Let's keep backward compatibility to a max at least till the next major release.

File: openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart (hunk 4 above)
Comment (Nexushunter) [reply openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:247]: Sorry for the slow reply. 🤔 I was tossing up between using a map vs the token string. One concern I have with placing the header generation on the user is the requirements that AWS has specifically. The request needs to be signed within a 15 minute window of being used and this will cause them to need to manually regenerate that header, thus requiring updates to the configuration. 

I may have a different use case where my flow would be to rarely update the Annotation but define the remote spec provider. While this does impose a bit on the library for the implementation it allows for the builder to properly handle the communication w/o requiring too much tweaking and manual changes from the developer. I personal would greatly dislike having to regenerate the signed header each and every time I want to regenerate the spec.

As for a  `GRPCRemoteSpec` I don't know enough about what is needed to even support the authentication story for that. 

Proposal:
- I change the RemoteSpec to take in headers
- We leave the AWS implementation so that the end user doesn't need to rebuild and sign the request
  - I can take ownership of the AWS implementation and thus when issues arise about it I will address the issues.
  - We only support the `GET` request to `S3` and we can take contributions for `GRPCRemoteSpec` or Azure etc with the explicit understanding they are responsible for both maintenance and triage. If it gets ignored by the contributor we will deprecate and drop support. 
    -  This may seem harsh but the support shouldn't need to fall on those that do not have an understanding of the subject matter.
    - This would require some level of staleness & community guidelines.
    - IIRC `GRPCRemoteSpec` would require that we have it installed into the repo and would require a larger lift due to the nature of GRPC, and this is likely unfeasible.

Comment (gibahjoe) [reply openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:247]: I don't mean the user will generate the header manually. 

What I mean is that they will provide the implementation that generates the header.

Let elaborate.

If `RemoteSpec` has a field `hearderDelegate` of type `OpenapiHeaderDelegate`. Now, this  `OpenapiHeaderDepegate` can be extended by the user for whatever remote source they use for their spec.

This of course does not prevent the library from providing a default implementation like `AWSHeaderDelegate` or `MapHeaderDelegate`. I think this is more maintainable. The last thing we want is dropping support for useful features without providing alternatives for existing users.

This is just an example. And the code makes sense in my head. lol. 

What do you think?

Comment (Nexushunter) [reply openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:247]: 🤔 That makes sense to me. I misunderstood before. 

Would you mind expanding a bit more on what the delegate might look like?

Comment (gibahjoe) [reply openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:247]: Sorry for late reply, been out of power. Something like 

```dart
///The header provider base class
class RemoteSpecHeaderDelegate{
  const RemoteSpecHeaderDelegate();
  Map<String,String>? getHeader(){
    return null;
  }
}

/// Default header provider provided by library
class AWSRemoteSpecHeaderDelegate extends RemoteSpecHeaderDelegate{
  
  Map<String,String>? getHeader(){
    return null;
  }

 const  AWSRemoteSpecHeaderDelegate();
}

/// header provider implemented by user
class UserImplementedHeaderDelegate extends RemoteSpecHeaderDelegate{

  const UserImplementedHeaderDelegate();

  @override
  Map<String, String>? getHeader() {
    return {'key':'value'};
  }
}

```

the below field is added to the openapi annotation or to the RemoteSpec class (preferrably). For example below, I added it to the openapi annotation

```dart
final RemoteSpecHeaderDelegate? headerDelegate;
```

and used like 

```dart
@Openapi(
  additionalProperties:
      DioProperties(pubName: 'petstore_api', pubAuthor: 'Johnny dep..'),
  inputSpecFile: 'openapi-spec.yaml',
  typeMappings: {'Pet': 'ExamplePet'},
  generatorName: Generator.dio,
  runSourceGenOnOutput: true,
  headerDelegate: UserImplementedHeaderDelegate(), // Header provider passed to the openapi annotation
  useNextGen: true,
  outputDirectory: 'api/petstore_api',
)
```

Comment (gibahjoe) [reply openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:247]: on a second thought, `headerProvider` is probably a better name or just `header`

Comment (gibahjoe) [reply openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:247]: I can try to explain better if you do not understand

File: openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart [hunk 5]
```diff
@@ -25,6 +31,19 @@ class Openapi {
   /// -i
   final String inputSpecFile;
```
Comment (gibahjoe) [inline openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:32]: With InputSpec, we can now deprecate this. It will be removed in version 5.0.0 (which will be released next week)

File: openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart [hunk 6]
```diff
@@ -152,6 +172,164 @@ class Openapi {
   });
 }
 
+/// Provides the input spec file to be used.
+///
+/// Provides the location of the input spec file to be used by the generator.
+/// Includes the option to use the default json or yaml paths.
+class InputSpec {
+  final String path;
+  final bool defaultYaml;
```
Comment (gibahjoe) [inline openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:181]: I don't think there is a need for this flag. You can just check for openapi.{yaml|yml|json} unless you are using it for something else

File: openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart [hunk 7]
```diff
@@ -152,6 +172,164 @@ class Openapi {
...
+
+  const InputSpec({String? path, this.defaultYaml = true})
+      : path = path ?? 'openapi.${defaultYaml ? 'yaml' : 'json'}';
```
Comment (gibahjoe) [inline openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:184]: you missed 'yml' which is a valid yaml file extension

**Review by gibahjoe — CHANGES_REQUESTED** [review]
Thank you for your contribution. Just a few feedback to attend to

File: openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart (hunk 6 above)
Comment (Nexushunter) [reply openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart:181]: I intended on it being used by the `GeneratorArguments` to provide the default openapi spec file. But it looks like I never made those changes. I can that tweaked.

**Review by gibahjoe — APPROVED** [review]
LGTM

**Comment by gibahjoe** [comment]
@Nexushunter , seems there is an issue with the workflow configuration. Please take a look so all checks can run
Screenshot 2023-09-02 at 15 51 41

**Comment by Nexushunter** [comment]
@gibahjoe I'll put up a fix in an hour. Looks like I messed up the merge from master sorry

**Comment by Nexushunter** [comment]
@gibahjoe It should be fixed now

**Review by gibahjoe — APPROVED** [review]
